### PR TITLE
add tags and loading state support

### PR DESCRIPTION
**Pull request by ticruz38**
- Adding a command getMetaTags that will return all imeta tags present in this extension.

- Adding a loading state, handy with immediateUpload where you want your file to be uploaded before allowing a submit.

- Recomputing the node position instead of using the pos parameter in updateNodeAttributes, when mixing this extension with other, the pos attribute sometimes was incorrect.

### CODE REVIEW

**Comment by cesardeazevedo** [comment]
I am no sure why we need getMetaTags or tags on storage function as I already put the tags on each file, concatenating all of them seems a not so useful abstraction

**Comment by ticruz38** [comment]
> I am no sure why we need getMetaTags or tags on storage function as I already put the tags on each file, concatenating all of them seems a not so useful abstraction

oh yes absolutely!

Actually I only used the storage tags, but then realised I could not track which are deleted and added this getMetaTags action. I'm using just that now

**Comment by cesardeazevedo** [comment]
I mess up and forgot to commit some error handling, can you rebase?

File: src/extensions/FileUploadExtension.ts [hunk 1]
```diff
@@ -75,11 +76,27 @@ export const FileUploadExtension = Extension.create<FileUploadOptions>({
         props.tr.setMeta('uploadFiles', true)
         return true
       },
+      getMetaTags: () => (props: CommandProps) => {
+        const tags: string[][] = []
+        // make sure the file uploaded is still in the editor content
+        props.editor.state.doc.descendants((node) => {
```
Comment (cesardeazevedo) [inline src/extensions/FileUploadExtension.ts:82]: I am not so sure on what you trying to do here, you are iterating on all documents looking for image and video nodes, when you already have them on `props.editor.storage.fileUpload.files`

**Comment by cesardeazevedo** [comment]
I am also not so sure about the `loading` flag, it doesn't help much the control flow as we still need to listen to the onComplete callback to make sure if all files were upload correctly, ideally the `uploadFiles` command would just return a promise but we can't, tiptap/prosemirror give us very little to no guidelines on how to properly do async stuff within the editor, since the `loading` flag can't be used as state on react or svelte, i think it's better for users control their own state by using the `onStart`, `onDrop` and `onComplete` callbacks, how you using the flag?

File: src/extensions/FileUploadExtension.ts (hunk 1 above)
Comment (ticruz38) [reply src/extensions/FileUploadExtension.ts:84]: The reason I am iterating here is to check if the file is still in the editor.

When you are using immediateUpload, maybe that file has been subsequently deleted, the deleted action is not tracked in the extension is it?

**Comment by ticruz38** [comment]
the onComplete trigger when using uploadFiles. 
If you are using selectFile and immediateUpload you would have no way to know if the upload is ongoing or not.
To be honest, the loading state is a svelte store in my use case. Here I can't make any assumption so loading can be useful fo you to cancel a submit if some files are still loading.

File: src/extensions/FileUploadExtension.ts (hunk 1 above)
Comment (cesardeazevedo) [reply src/extensions/FileUploadExtension.ts:84]: You are correct, I don't think we should manually populate the storage files as we can't track when the user deletes an image, we should follow the same approach of NostrExtension storage get functions with something like `getFiles()` https://github.com/cesardeazevedo/nostr-editor/blob/main/src/extensions/NostrExtension.ts#L139-L182

**Comment by cesardeazevedo** [comment]
I just got some interesting improvements here.

I moved the Uploader instance to the storage, then we could trigger a upload and receive a proper promise, and also have more flexibility on access code on the Uploader class.

```ts
const handleUpload = async () => {
  try {
    const files = await editor.storage.fileUpload.uploader.start()
  } catch (error) {
    console.log(error)
  }
}
```

You were right about the onComplete, It wasn't working with immediateUpload true, I'm trying to get it all working at once.

My apologies for making more conflicts in this pull-request, I would suggest you to not push any further changes until I push this, sorry.

**Comment by cesardeazevedo** [comment]
Just pushed new improvements and fixes.

* `onComplete` now should trigger on both `uploadFiles` and `immediateUpload` true.
* Now you can trigger upload with `await editor.storage.fileUpload.uploader.start()` which returns a promise with the node attributes, but it will throw an error if any file were rejected.
* I removed `files` from storage in favor of `getFiles()` which get's you the current state of the files in the editor.

let me know what you think.

**Comment by ticruz38** [comment]
yes this should cover all use cases

**Comment by staab** [comment]
> For another PR, having all the tags present in the current editor content from a single command would be amazing too. ie nprofile, imeta tags, all sort of tags to be added in a Nostr event.

@cesardeazevedo we have some code you can borrow if you want to: https://github.com/coracle-social/flotilla/blob/master/src/lib/editor/util.ts

**Comment by cesardeazevedo** [comment]
Thanks, just added `getEditorTags()` function. unit tests can be found here.
https://github.com/cesardeazevedo/nostr-editor/blob/main/src/__tests__/storage.test.ts#L59-L102

Hope it's correct

**Comment by staab** [comment]
I think opinions differ on whether to refer to non-replaceable events using an address, my recommendation would be to do that only if the event is replaceable. I also think the `a` should probably be a `q`, otherwise the note will show up as a reply.

**Comment by cesardeazevedo** [comment]
The `a` tag is from the naddr embeds, nevent embeds results in `q` tags, I agree naddress only makes sense for replaceable events, I embed a naddress from a kind 1 event on my tests for simplicity only

**Comment by staab** [comment]
Ah, ok I see. I think my second comment still applies.

**Comment by cesardeazevedo** [comment]
Hm, not so sure if I understood, how can we embed naddress as `q` tags? or it's possible to do `["q", "30023:<pubkey>:<identifier>]`?

**Comment by staab** [comment]
I don't think there's a convention for it right now. Flotilla uses a q tag of the same form as the event id-based version. It might be more correct based on the current spec (which I edited a week or two ago) to convert the address to an event id for the `q` tag, but in some situations that may not be possible. So I think `["q", "address", "relay"]` is probably the way to go.

**Pull request closed without merging**